### ISSUE TITLE
[Bug]: _upsample_nearest_exact2d_backward : TritonShared RVV lowering fails on tl.atomic_add / tt.atomic_rmw before VectorIR

### ISSUE THREAD

**Issue by 111a909** [RuyiAI-Stack/triton-riscv#52]
### Triton python code

```mlir
import os

os.environ["TRITON_BACKENDS_IN_TREE"] = "1"
_SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
_REPO_DIR = os.path.abspath(os.path.join(_SCRIPT_DIR, "..", ".."))
_TRITON_HOME = os.path.join(_REPO_DIR, ".triton_home")
os.environ.setdefault("TRITON_HOME", _TRITON_HOME)
os.environ.setdefault("TRITON_CACHE_DIR", os.path.join(_TRITON_HOME, "cache"))
os.environ.setdefault("TRITON_DUMP_DIR", os.path.join(_TRITON_HOME, "dump"))
os.environ.setdefault("TRITON_OVERRIDE_DIR", os.path.join(_TRITON_HOME, "override"))
os.environ["TRITON_SHARED_DUMP_PATH"] = os.path.join(_TRITON_HOME, "dump", "shared")
for _path in (
    os.environ["TRITON_CACHE_DIR"],
    os.environ["TRITON_DUMP_DIR"],
    os.environ["TRITON_OVERRIDE_DIR"],
    os.environ["TRITON_SHARED_DUMP_PATH"],
):
    os.makedirs(_path, exist_ok=True)

try:
    import torch
except ModuleNotFoundError:
    torch = None

import triton
import triton.language as tl
from triton.compiler.compiler import ASTSource, GPUTarget


@triton.jit
def _upsample_nearest_exact2d_backward_kernel(
    grad_output_ptr,
    grad_input_ptr,
    N,
    C,
    OH,
    OW,
    IH,
    IW,
    reciprocal_scale_h: tl.constexpr,
    reciprocal_scale_w: tl.constexpr,
    BLOCK_SIZE: tl.constexpr,
):
    # Get program IDs
    pid = tl.program_id(0)
    nc_pid = tl.program_id(1)

    # Each thread handles one output position
    # Total output positions = OH * OW
    idx = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)

    # Filter valid indices
    mask = idx < (OH * OW)

    # Compute oh, ow from linear index
    ow = idx % OW
    oh = idx // OW

    # Compute corresponding input indices
    ih = (oh * reciprocal_scale_h).to(tl.int32)
    iw = (ow * reciprocal_scale_w).to(tl.int32)

    # Clamp to valid range
    ih = tl.minimum(ih, IH - 1)
    iw = tl.minimum(iw, IW - 1)

    # Compute offsets
    # For output: (nc, oh, ow) -> linear
    # For input: (nc, ih, iw) -> linear
    offset_o = (nc_pid * OH + oh) * OW + ow
    offset_i = (nc_pid * IH + ih) * IW + iw

    # Load gradient and accumulate
    grad = tl.load(grad_output_ptr + offset_o, mask=mask)
    tl.atomic_add(grad_input_ptr + offset_i, grad, mask=mask)


def _upsample_nearest_exact2d_backward(
    grad_output,
    output_size,
    input_size,
    scales_h=None,
    scales_w=None,
):
    if torch is None:
        raise RuntimeError("torch is required for Python-level execution")

    assert grad_output.ndim == 4, "The ndim of grad_output must be 4"
    assert len(output_size) == 2, "The len of output_size must be 2"
    assert len(input_size) == 4, "The len of input_size must be 4"

    OH, OW = output_size
    N, C, IH, IW = input_size

    assert grad_output.shape == (N, C, OH, OW), (
        f"grad_output shape {grad_output.shape} does not match "
        f"expected shape (N={N}, C={C}, OH={OH}, OW={OW})"
    )

    if scales_h is not None:
        reciprocal_scale_h = 1 / scales_h
    else:
        reciprocal_scale_h = IH / OH
    if scales_w is not None:
        reciprocal_scale_w = 1 / scales_w
    else:
        reciprocal_scale_w = IW / OW

    grad_input = torch.zeros(
        (N, C, IH, IW), device=grad_output.device, dtype=torch.float32
    )

    BLOCK_SIZE = 128
    grid = (triton.cdiv(OH * OW, BLOCK_SIZE), N * C)

    _upsample_nearest_exact2d_backward_kernel[grid](
        grad_output,
        grad_input,
        N,
        C,
        OH,
        OW,
        IH,
        IW,
        reciprocal_scale_h,
        reciprocal_scale_w,
        BLOCK_SIZE,
    )

    return grad_input.to(grad_output.dtype)


def _rename_kernel_symbol(ir, name):
    return ir.replace("@_upsample_nearest_exact2d_backward_kernel", f"@{name}")


def _export_case(name, reciprocal_scale_h, reciprocal_scale_w):
    src = ASTSource(
        fn=_upsample_nearest_exact2d_backward_kernel,
        signature={
            "grad_output_ptr": "*fp32",
            "grad_input_ptr": "*fp32",
            "N": "i32",
            "C": "i32",
            "OH": "i32",
            "OW": "i32",
            "IH": "i32",
            "IW": "i32",
            "reciprocal_scale_h": "constexpr",
            "reciprocal_scale_w": "constexpr",
            "BLOCK_SIZE": "constexpr",
        },
        constexprs={
            "reciprocal_scale_h": reciprocal_scale_h,
            "reciprocal_scale_w": reciprocal_scale_w,
            "BLOCK_SIZE": 128,
        },
    )
    ret = triton.compile(src, target=GPUTarget("cpu", 0, 0))
    for stage, ext in [("ttir", "ttir"), ("llir", "ll")]:
        path = f"output/{name}.{ext}"
        text = ret.asm[stage]
        if stage == "llir":
            text = _rename_kernel_symbol(text, name)
        with open(path, "w") as f:
            f.write(text)
        print(f"OK {path} ({len(text)} bytes)")


def export_qemu_cases():
    os.makedirs("output", exist_ok=True)
    _export_case(
        "upsample_nearest_exact2d_backward_size",
        reciprocal_scale_h=3.0 / 5.0,
        reciprocal_scale_w=4.0 / 7.0,
    )
    _export_case(
        "upsample_nearest_exact2d_backward_scale",
        reciprocal_scale_h=2.0 / 3.0,
        reciprocal_scale_w=4.0 / 5.0,
    )


if __name__ == "__main__":
    export_qemu_cases()
```

### Triton IR

```mlir
#loc = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":31:1)
#loc26 = loc("grad_output_ptr"(#loc))
#loc27 = loc("grad_input_ptr"(#loc))
#loc28 = loc("N"(#loc))
#loc29 = loc("C"(#loc))
#loc30 = loc("OH"(#loc))
#loc31 = loc("OW"(#loc))
#loc32 = loc("IH"(#loc))
#loc33 = loc("IW"(#loc))
module {
  tt.func public @_upsample_nearest_exact2d_backward_kernel(%grad_output_ptr: !tt.ptr<f32> loc("grad_output_ptr"(#loc)), %grad_input_ptr: !tt.ptr<f32> loc("grad_input_ptr"(#loc)), %N: i32 loc("N"(#loc)), %C: i32 loc("C"(#loc)), %OH: i32 loc("OH"(#loc)), %OW: i32 loc("OW"(#loc)), %IH: i32 loc("IH"(#loc)), %IW: i32 loc("IW"(#loc))) attributes {noinline = false} {
    %c1_i32 = arith.constant 1 : i32 loc(#loc1)
    %iw = arith.constant dense<0.571428597> : tensor<128xf32> loc(#loc34)
    %ih = arith.constant dense<6.000000e-01> : tensor<128xf32> loc(#loc35)
    %c128_i32 = arith.constant 128 : i32 loc(#loc1)
    %pid = tt.get_program_id x : i32 loc(#loc36)
    %nc_pid = tt.get_program_id y : i32 loc(#loc37)
    %idx = arith.muli %pid, %c128_i32 : i32 loc(#loc38)
    %idx_0 = tt.make_range {end = 128 : i32, start = 0 : i32} : tensor<128xi32> loc(#loc39)
    %idx_1 = tt.splat %idx : i32 -> tensor<128xi32> loc(#loc38)
    %idx_2 = arith.addi %idx_1, %idx_0 : tensor<128xi32> loc(#loc38)
    %mask = arith.muli %OH, %OW : i32 loc(#loc40)
    %mask_3 = tt.splat %mask : i32 -> tensor<128xi32> loc(#loc41)
    %mask_4 = arith.cmpi slt, %idx_2, %mask_3 : tensor<128xi32> loc(#loc41)
    %ow = tt.splat %OW : i32 -> tensor<128xi32> loc(#loc42)
    %ow_5 = arith.remsi %idx_2, %ow : tensor<128xi32> loc(#loc42)
    %oh = arith.divsi %idx_2, %ow : tensor<128xi32> loc(#loc43)
    %ih_6 = arith.sitofp %oh : tensor<128xi32> to tensor<128xf32> loc(#loc35)
    %ih_7 = arith.mulf %ih_6, %ih : tensor<128xf32> loc(#loc35)
    %ih_8 = arith.fptosi %ih_7 : tensor<128xf32> to tensor<128xi32> loc(#loc44)
    %iw_9 = arith.sitofp %ow_5 : tensor<128xi32> to tensor<128xf32> loc(#loc34)
    %iw_10 = arith.mulf %iw_9, %iw : tensor<128xf32> loc(#loc34)
    %iw_11 = arith.fptosi %iw_10 : tensor<128xf32> to tensor<128xi32> loc(#loc45)
    %ih_12 = arith.subi %IH, %c1_i32 : i32 loc(#loc46)
    %ih_13 = tt.splat %ih_12 : i32 -> tensor<128xi32> loc(#loc47)
    %ih_14 = arith.minsi %ih_8, %ih_13 : tensor<128xi32> loc(#loc47)
    %iw_15 = arith.subi %IW, %c1_i32 : i32 loc(#loc48)
    %iw_16 = tt.splat %iw_15 : i32 -> tensor<128xi32> loc(#loc49)
    %iw_17 = arith.minsi %iw_11, %iw_16 : tensor<128xi32> loc(#loc49)
    %offset_o = arith.muli %nc_pid, %OH : i32 loc(#loc50)
    %offset_o_18 = tt.splat %offset_o : i32 -> tensor<128xi32> loc(#loc50)
    %offset_o_19 = arith.addi %offset_o_18, %oh : tensor<128xi32> loc(#loc50)
    %offset_o_20 = arith.muli %offset_o_19, %ow : tensor<128xi32> loc(#loc51)
    %offset_o_21 = arith.addi %offset_o_20, %ow_5 : tensor<128xi32> loc(#loc51)
    %offset_i = arith.muli %nc_pid, %IH : i32 loc(#loc52)
    %offset_i_22 = tt.splat %offset_i : i32 -> tensor<128xi32> loc(#loc52)
    %offset_i_23 = arith.addi %offset_i_22, %ih_14 : tensor<128xi32> loc(#loc52)
    %offset_i_24 = tt.splat %IW : i32 -> tensor<128xi32> loc(#loc53)
    %offset_i_25 = arith.muli %offset_i_23, %offset_i_24 : tensor<128xi32> loc(#loc53)
    %offset_i_26 = arith.addi %offset_i_25, %iw_17 : tensor<128xi32> loc(#loc53)
    %grad = tt.splat %grad_output_ptr : !tt.ptr<f32> -> tensor<128x!tt.ptr<f32>> loc(#loc54)
    %grad_27 = tt.addptr %grad, %offset_o_21 : tensor<128x!tt.ptr<f32>>, tensor<128xi32> loc(#loc54)
    %grad_28 = tt.load %grad_27, %mask_4 : tensor<128x!tt.ptr<f32>> loc(#loc55)
    %0 = tt.splat %grad_input_ptr : !tt.ptr<f32> -> tensor<128x!tt.ptr<f32>> loc(#loc24)
    %1 = tt.addptr %0, %offset_i_26 : tensor<128x!tt.ptr<f32>>, tensor<128xi32> loc(#loc24)
    %2 = tt.atomic_rmw fadd, acq_rel, gpu, %1, %grad_28, %mask_4 : (tensor<128x!tt.ptr<f32>>, tensor<128xf32>, tensor<128xi1>) -> tensor<128xf32> loc(#loc25)
    tt.return loc(#loc)
  } loc(#loc)
} loc(#loc)
#loc1 = loc(unknown)
#loc2 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":61:11)
#loc3 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":60:11)
#loc4 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":45:11)
#loc5 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":46:14)
#loc6 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":50:11)
#loc7 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":50:30)
#loc8 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":53:19)
#loc9 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":53:12)
#loc10 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":56:10)
#loc11 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":57:10)
#loc12 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":60:10)
#loc13 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":61:10)
#loc14 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":64:25)
#loc15 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":64:10)
#loc16 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":65:25)
#loc17 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":65:10)
#loc18 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":70:17)
#loc19 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":70:16)
#loc20 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":71:17)
#loc21 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":71:16)
#loc22 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":74:20)
#loc23 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":74:12)
#loc24 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":75:19)
#loc25 = loc("/home/dingbl/triton-riscv-env/rvv_test_qemu/upsample_nearest_exact2d_backward/upsample_nearest_exact2d_backward.py":75:5)
#loc34 = loc("iw"(#loc2))
#loc35 = loc("ih"(#loc3))
#loc36 = loc("pid"(#loc4))
#loc37 = loc("nc_pid"(#loc5))
#loc38 = loc("idx"(#loc6))
#loc39 = loc("idx"(#loc7))
#loc40 = loc("mask"(#loc8))
#loc41 = loc("mask"(#loc9))
#loc42 = loc("ow"(#loc10))
#loc43 = loc("oh"(#loc11))
#loc44 = loc("ih"(#loc12))
#loc45 = loc("iw"(#loc13))
#loc46 = loc("ih"(#loc14))
#loc47 = loc("ih"(#loc15))
#loc48 = loc("iw"(#loc16))
#loc49 = loc("iw"(#loc17))
#loc50 = loc("offset_o"(#loc18))
#loc51 = loc("offset_o"(#loc19))
#loc52 = loc("offset_i"(#loc20))
#loc53 = loc("offset_i"(#loc21))
#loc54 = loc("grad"(#loc22))
#loc55 = loc("grad"(#loc23))
```

### Crash log

```shell
error: unexpected op in ptr sequence
note: see current operation: "tt.atomic_rmw"(...)
```

### Additional information

This is not a QEMU/runtime failure; lowering stops before LLVM IR because `tl.atomic_add` remains as `tt.atomic_rmw`, which `triton-to-unstructured` does not legalize and `buddy-opt --one-shot-bufferize` cannot bufferize.

**Comment by maochuanjun** [comment]
Thanks for the detailed report. This has been fixed on the current main branch by #44.

  The previous pipeline did not recognize `tt.atomic_rmw` while rewriting unstructured pointer sequences. The updated lowering now converts:

  `tt.atomic_rmw` → `tts.atomic_rmw` → `memref.atomic_rmw`

  while preserving the mask, atomic operation, memory semantics, and scope.

  I replayed the exact TTIR attached to this issue on the current main branch. It successfully passed TTIR-to-TTShared lowering, one-shot bufferization, VectorIR generation, and LLVM IR generation. No `tt.atomic_rmw` remains in the generated VectorIR.

  Please update to the latest main branch and rebuild the compiler. This issue can be closed if it no longer reproduces in the reporter's environment.
